src/protections/MentionFlood.ts

```diff
@@ -0,0 +1,116 @@
+
+    settings = {};
+
+    private justJoined: { [roomId: string]: { [username: string]: Date; }; } = {};
+    private mention: RegExp;
+
+    constructor() {
+        this.mention = new RegExp(`@${LOCALPART_REGEX}:(${DOMAIN_REGEX}|${IPV4_REGEX}|${IPV6_REGEX})${PORT_REGEX}`, "g");
+    }
+
+    public get name(): string {
+        return 'MentionFlood';
+    }
+
+    public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
+
+        const content = event['content'] || {};
+        const minsBeforeTrusting = config.protections.mentionFlood.minutesBeforeTrusting;
+
+        if (minsBeforeTrusting > 0) {
+            if (!this.justJoined[roomId]) this.justJoined[roomId] = {};
+
+            // When a new member logs in, store the time they joined.  This will be useful
+            // when we need to check if a message was sent within 20 minutes of joining
+            if (event['type'] === 'm.room.member') {
+                if (isTrueJoinEvent(event)) {
+                    const now = new Date();
+                    this.justJoined[roomId][event['state_key']] = now;
+                    LogService.info("MentionFlood", `${event['state_key']} joined ${roomId} at ${now.toDateString()}`);
+                } else if (content['membership'] === 'leave' || content['membership'] === 'ban') {
+                    delete this.justJoined[roomId][event['sender']];
+                }
+
+                return;
+            }
+        }
+
+        if (event['type'] === 'm.room.message') {
+            const message = content['formatted_body'] || content['body'] || null;
+
+            // Check conditions first
+            if (minsBeforeTrusting > 0) {
+                const joinTime = this.justJoined[roomId][event['sender']];
+                if (joinTime) { // Disregard if the user isn't recently joined
+
+                    // Check if they did join recently, was it within the timeframe
+                    const now = new Date();
+                    if (now.valueOf() - joinTime.valueOf() > minsBeforeTrusting * 60 * 1000) {
+                        delete this.justJoined[roomId][event['sender']]; // Remove the user
+                        LogService.info("MentionFlood", `${event['sender']} is no longer considered suspect`);
+                        return;
+                    }
+
+                } else {
+                    // The user isn't in the recently joined users list, no need to keep
+                    // looking
+                    return;
+                }
+            }
+
+
+            // Perform the test
+            const maxMentionsPerMessage = config.protections.mentionFlood.maxMentionsPerMessage;
+            if (message && message.match(this.mention).length > maxMentionsPerMessage) {
+                await logMessage(LogLevel.WARN, "MentionFlood", `Banning ${event['sender']} for mention flood violation in ${roomId}.`);
+                if (!config.noop) {
```

Comment:
This also should move to `settings`. Also, it would be much better if we could have a choice between `kick`, `ban`, `warn` (to do nothing).

Reply:
Hm where should the warning actually happen? Warning the moderator or warning the user?

Reply:
Warning the moderator.